### PR TITLE
Bugfix: Prevent Dropdown from calling SelectContext function of surrounding components (#4176)

### PR DESCRIPTION
**Pull request by BenStadnick**
This pull request is to fix the bug described in issue #4176.

When a DropdownItem is selected, the Dropdown component calls a function provided by SelectableContext. This allows a nested Dropdown to close parent Dropdowns but also causes Dropdown to call functions provided by other components SelectableContext as well, e.g. an Accordion component.

This fix is internal to Dropdown and has no API changes. It includes a unit tests to check that Dropdown does not call functions provided by SelectableContext unless they come from a parent Dropdown. It also includes a unit test to check that selecting a DropdownItem in a nested Dropdown will close the parent Dropdowns as well.

### CODE REVIEW

File: src/Dropdown.js [hunk 1]
```diff
@@ -112,10 +112,11 @@ const Dropdown = React.forwardRef((uncontrolledProps, ref) => {
   );
 
   const handleSelect = useEventCallback((key, event) => {
-    if (onSelectCtx) onSelectCtx(key, event);
+    if (onSelectCtx && onSelectCtx.isDropdownSelCtx) onSelectCtx(key, event);
     if (onSelect) onSelect(key, event);
     handleToggle(false, event, 'select');
   });
+  handleSelect.isDropdownSelCtx = true;
```
Comment (taion) [inline src/Dropdown.js]: how does mutating the result of `useEventCallback` do anything to the value of `useContext(SelectableContext)` above?

Comment (taion) [reply src/Dropdown.js]: in other words you're just hard-disabling this functionality, which is wrong

Comment (BenStadnick) [reply src/Dropdown.js]: The idea is to modify the the context it provides to child components via
```JavaScript
<SelectableContext.Provider value={handleSelect}>
```
This way onSelectCtx will only fire if the parent is a Dropdown, closing that parent. I checked and this fix does break NavDropdown, so I will try another approach. I also found that selecting a Nav.Item will cause an Accordion to collapse as well. So perhaps Accordion would be a better place to fix this issue.

I still think a unit test for nested Dropdown behavior is good to have as this behavior is provided by bootstrap via the dropdown-submenu class.

Comment (taion) [reply src/Dropdown.js]: you're mutating the return value of a `useEventCallback` hook. and you shouldn't be mutating used context anyway

**Comment by taion** [comment]
did you see my PR comment above?

the point of this logic was to make `NavDropdown` work correctly. the right fix is to move the consumption of the selectable context to `NavDropdown`

**Comment by BenStadnick** [comment]
I’d like to stop the Dropdown from collapsing the Accordion as in issue #4176. Moving consumption to NavDropdown would accomplish this. However, other components which consume SelectContext, such as AbstractNav, also collapse a parent Accordion. For example, selecting an item in Nav will collapse the accordion in the example code bellow:
```JavaScript
import React from "react";
import { Nav, Card, Accordion } from "react-bootstrap";
const Test = props => {
  return (
    <Accordion>
      <Card>
        <Card.Header>
          <Accordion.Toggle eventKey="0">Toggle 1</Accordion.Toggle>
        </Card.Header>
        <Accordion.Collapse eventKey="0">
          <Card.Body>
            <Nav activeKey="1">
              <Nav.Item>
                <Nav.Link eventKey="1">
                  NavLink 1 content
                </Nav.Link>
              </Nav.Item>
              <Nav.Item>
                <Nav.Link eventKey="2">
                  NavLink 2 content
                </Nav.Link>
              </Nav.Item>
            </Nav>
          </Card.Body>
        </Accordion.Collapse>
      </Card>
    </Accordion>
  );
}

export default Test
```
TabPane blocks SelectContext in a similar way for nested Nav components. Excerpt from line 168-177:
```JavaScript
  // We provide an empty the TabContext so `<Nav>`s in `<TabPane>`s don't
  // conflict with the top level one.
  return (
    <TabContext.Provider value={null}>
      <SelectableContext.Provider value={null}>
        {pane}
      </SelectableContext.Provider>
    </TabContext.Provider>
  );
});
```

Moving consumption to NavDropdown will solve the Dropdown case but it would be good to handle other cases as well.

**Comment by taion** [comment]
abstract nav provides its own selectable context, so it wouldn't interact with the parent accordion

**Comment by BenStadnick** [comment]
Yes, AbstractNav provides context to its children but it also grabs the surrounding context, line 49,
```JavaScript
const parentOnSelect = useContext(SelectableContext);
```
When I run the code posted above, selecting an item in the Nav collapses the accordion.

**Comment by kyletsang** [comment]
Fixed in #5201

**Pull request closed without merging**